### DISCUSSION TITLE
Experiences with new color correction profile settings?

### DISCUSSION THREAD

**Discussion by Jottle** [profi200/open_agb_firm#192] [General]
I just enabled the gba color correciton profile on my N3ds xl with IPS screens. I used the recommended gamma of 1.8, but even with the brightness all the way up, the screen still looks kind of dark/muddy compared to the screenshots of the profile in action [here](https://i.ibb.co/6Z8tx29/all-sbs.png). 

I'm sure this is due to variations in calibration between units, but what can I change to make the color profile look a big brighter without sacrificing color accuracy? Should I bump the gamma number up to 1.9 or 2.0? Or back down?

Anyone else have good results?

**Comment by dLeandroT** [comment]
I set the `lcdGamma` to 1.44 on TN screen. With the recommended settings, it was too bright. 
You can try increasing the `lcdGamma` to enhance the brightness. Try `lcdGamma=2.2`.

In fact, in the FAQ section, `2.2` is the recommended value for making the colors appear lighter."

**Reply by Jottle** [reply]
Oh interesting. I guess I read the information as conflicting b/c the gamma 2.2 recommendation is for when you're not using the gba color correction profile, just the default 3DS LCD profile. I'll split the difference (a bit) and try playing with 2.0 or 2.1. for my IPS.

**Reply by dLeandroT** [reply]
I mean, that parameter seems to be the recommended way to adjust the screen.
I also have to say that I compared it with my AGB 001.

**Comment by profi200** [comment]
The gamma curve of these LCDs seems to be what's most off and also i don't know if your laptop/PC monitor is calibrated. So there are many variables. 1.8 seems to be roughly right to me. I compared to my own GBA SP using a gradient from black to white which shows the gamma curve. Also note that N3DS XL IPS LCDs crush blacks very noticeably. Near black most color bars pretty much disappear.

https://www.neogaf.com/threads/detailed-review-of-switch-screen-quality-and-compared-to-3ds-by-erica-griffin.1369087/

**Reply by Jottle** [reply]
I agree that's it's probably some combination of how my screen reproduces the example images using the gba color profile and variations in the 3ds screens out there. It's a tradeoff between accuracy vs. preference really since the 3ds screen just can't exactly match the gba color profile. But I guess getting it closer to the original is better than nothing.

**Comment by profi200** [comment]
I might have a fix for the gamma issue. Turns out we know the gamma curves for certain LCDs Nintendo uses and can apply some small correction to get it to match gamma 2.2 roughly.

**Reply by Jottle** [reply]
Amazing! So we would just need to know whether our top screen is ips/tn? Could we implement it in the config.ini? Or would we need to wait until the next beta? Thanks @profi200!

**Reply by profi200** [reply]
Unfortunately no. This curve is not a standard gamma curve but has a distinct shape (see below) we can't represent with gbaGamma or lcdGamma. According to my tests and what others tested this same curve seems to match for other LCDs as well. Only problem is that color temperatures vary wildly with each. Literally every 3DS has different color temperatures.

Top RGB lines show the deviation from standard gamma 2.2 and the bottom lines show what i want to do.
![3ds_top_lcd_gamma](https://github.com/user-attachments/assets/1cd73269-0b1a-4b68-bb7a-1d602d19211b)

**Reply by Jottle** [reply]
So surprising that these screens vary so widely. Will you just average these for the next update?

**Reply by profi200** [reply]
Not sure what you mean with average. My plan is to cancel the deviation out almost completely by using the hardware gamma table. As can be seen in that diagram there are duplicated values because this table uses integers so the fractions are lost (accurate adjustments can only be done on the LCD side).

**Comment by Gregoras** [comment]
I have New 3DS XL with upper IPS screen. For me the biggest issue is not gamma, but colors temperature. On my unit colors look to cold, rather blueish. Hardly to even talking about proper brightness with that issue.

**Reply by profi200** [reply]
The fix for gamma i have planned doesn't fix the color temperature. Honestly i don't actually know how i can compute the curves as seen above in the picture while also applying redshift.

**Reply by Gregoras** [reply]
I don't expect that you are able fixing any problem which Nintendo not even notice or care. Even if you abandon the project right now, OpenAGB will be still awesome.
I'm just crying that I'm cannot even complain about to bright or to dimmer screen. Colors always looking off and hardly to tell if increasing or decreasing brightness makes things better or worse.

**Reply by TuxSH** [reply]
> On my unit colors look to cold, rather blueish

This depends on your unit, some of the units have a warm color temperature while others have colder temperature. Same for my DSiXLs. I advise you to check with games that have white screen transitions, like Mario Kart 7.

**Reply by profi200** [reply]
Yes, this is basically Nintendos screen lottery everyone is talking about. Different LCDs, different color temperatures ect. The later seems to also be a problem on the Switch. Some say they got cool looking LCD panels and others got overly warm ones almost looking yellow.

**Reply by Gregoras** [reply]
But wasn't the same at GBA time?

**Reply by profi200** [reply]
They only got 2 LCDs as far as i know. The AGB-/AGS-001 one (dark, even darker than Game Boy Color) and AGS-101 which is similar to a DS lite. Manufacturing tolerances are also there but i think it's more consistent than the mess with backlit consoles.

**Comment by profi200** [comment]
How is that build? Note that the gbaGamma, lcdGamma, brightness and contrast settings don't do anything in this build. It should noticeably improve the colors (with and without color correction) and not crush the details.

[open_agb_firm_curve_correction.zip](https://github.com/user-attachments/files/16415202/open_agb_firm_curve_correction.zip)
Changes: https://gist.github.com/profi200/d26e8033c2ee43b65a6ad514d4b24cf4

**Comment by Contra-S** [comment]
I've tried different values and it just looks desaturated to me. Golden Sun looks brownish, for example. It must be a hardware issue, but maybe the color could be saturated a little to fix it?

**Reply by profi200** [reply]
The desaturated look is unfortunately expected. I think they have not figured out the perfect settings yet and also the shader this is based on mimics the GBA LCD under direct sunlight as far as i understand.

You can try the build i linked 1 message above yours. It should improve the gamma no matter if you use color correction or not.

**Reply by Jottle** [reply]
If we have ipa, should we still use gamma at 1.8 for this build?

**Reply by profi200** [reply]
Doesn't matter. These settings are ignored with that build.

**Reply by Contra-S** [reply]
I tried the new build. It looks similar to 1.54 lcdGamma, but lighter; kind of like 2.2, but without looking too washed out.

I find the colors more problematic than the gamma, but I guess there's not much that can be done about that.

Here's a crappy picture of how the build looks compared to mGBA with GBA Color on:
![2024-07-29-17-56-06-642](https://github.com/user-attachments/assets/83849826-a16d-49fb-ac9c-b53a704156d1)

And here one of only the 3DS display, because it looks blue on the upper one for some reason:
![2024-07-29-18-14-52-265](https://github.com/user-attachments/assets/9c38e8fb-6c9c-4248-9540-6a31b5a668f5)

**Reply by profi200** [reply]
Note that mGBA's GBA color shader is not up to date. It's an older variant of Pokefan's shader. It looks a bit different and more saturated. Maybe i will make my own shader specially designed for the 3DS that mimics the more saturated colors on a calibrated AGS-001. Also note that the color temperature of a 3DS LCD is apparently 6644K which is quite cool.

Also fwiw correcting the gamma solved one of the biggest issue.

**Comment by profi200** [comment]
Added a new release that has the new gamma curves implemented.

https://github.com/profi200/open_agb_firm/releases

**Reply by Jottle** [reply]
Thanks so much. Do we need to change our config.ini to take advantage of this change? Or it's built in with no need to change the config file? And if we are using custom gba color correction profiles in config.ini, this new beta will just ignore them, correct?

**Reply by profi200** [reply]
See the release description. Currently all settings except color profile are ignored.

**Comment by profi200** [comment]
I should maybe post this here too. The 3DS LCDs are even worse than i expected. They fail pretty much **_ALL_** tests with large error rates.

https://forums.libretro.com/t/real-gba-and-ds-phat-colors/1540/269

**Reply by Jottle** [reply]
Is this why gamma correction isn't in the latest release?

**Reply by profi200** [reply]
There is already a hardcoded gamma table that is meant to fix this but i have not verified if it does correct the curves correctly. That being said the gamma settings will come back at some point.
